src/memory_manager.rs

```diff
@@ -248,37 +313,82 @@ impl<M: Memory> MemoryManagerInner<M> {
             memory_sizes_in_pages: [0; MAX_NUM_MEMORIES as usize],
             memory_buckets: BTreeMap::new(),
             bucket_size_in_pages,
+            unallocated_buckets: {
+                let mut list = LinkedList::new();
+                for i in 0..MAX_NUM_BUCKETS {
+                    list.push_back(BucketId(i as u16));
+                }
+                list
+            },
```

Comment:
It seems a bit wasteful that we create a list of 32k elements here. How about, instead of tracking all of the `unallocated_buckets`, we only track the `freed_buckets`? When a memory needs to grow, we first use the buckets in the `freed_buckets`, and if that's empty, we allocate new buckets, as we do in V1. WDYT?